### PR TITLE
Add handling of parameter references in Sphinx documentation

### PR DESCRIPTION
**Pull request by klecki**
## Description:
Utilize [sphinx_paramlinks ](https://github.com/sqlalchemyorg/sphinx-paramlinks) plugin which adds:
* link target for every parameter
* new :paramref: directive rendering the parameter as a link (monospaced highlight)

Add hook that automatically injects :paramref: before every single-backticked parameter reference.

The references are validated against function signature. If the signature is unavaliable than the whole step is skipped.

Fix the numpydoc generation - no single ticks are required in the parameter documentation sections.

TODO: 
- [ ] some debug prints to be removed
- [x] install the plugin in the doc-build env.
- [x] detickyfy the handwritten docs (like external source)

### Example
random_bbox_crop documentation source:
dali/operators/image/crop/bbox_crop.cc:175
```
The third output contains the bounding boxes, after filtering them out by centroid or area thresholding
(see `bbox_prune_threshold` argument), and with the coordinates mapped to the new coordinate space.
```
Result:
obraz
(The link works)

Or:
obraz

### Affected modules and functionalities:
Docs, sphinx

### Tests:

- [ ] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [x] N/A
Just take a look at the docs

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: qa/setup_packages.py [hunk 1]
```diff
@@ -388,13 +388,13 @@ class CudaHttpPackage(CudaPackage):
...
-    `key`: str
+    key : str
         Name this package should be queried for
-    `versions`: str or PckgVer class object
+    versions : str or PckgVer class object
         Dictionary, where the key is a cuda version and vale, is the list
         of directly accessible http links. `{platform}` inside the link is replaced by the
         compatible platform tag provided by pip
-    `name`: str, , optional, default = None
+    name : str, , optional, default = None
```
Comment (mzient) [inline qa/setup_packages.py:397]:
```suggestion
name : str, optional, default = None
```

File: qa/setup_packages.py [hunk 2]
```diff
@@ -265,11 +265,11 @@ class PlainPackage(BasePackage):
...
     Parameters
     ----------
-    `key`: str
+    key : str
         Name this package should be queried for
-    `versions`: str or PckgVer class object
+    versions : str or PckgVer class object
         List of versions this package is available for
-    `name`: str, , optional, default = None
+    name : str, , optional, default = None
```
Comment (mzient) [inline qa/setup_packages.py:272]:
```suggestion
    name : str, optional, default = None
```

File: qa/setup_packages.py [hunk 3]
```diff
@@ -286,11 +286,11 @@ class CudaPackage(BasePackage):
...
     Parameters
     ----------
-    `key`: str
+    key : str
         Name this package should be queried for
-    `versions`: dict or PckgVer class object
+    versions : dict or PckgVer class object
         Dictionary, where the key is a cuda version and vale, is the list of versions supported
-    `name`: str, , optional, default = None
+    name : str, , optional, default = None
```
Comment (mzient) [inline qa/setup_packages.py:293]:
```suggestion
    name : str, optional, default = None
```

**Review by szkarpinski — APPROVED** [review]

**Review by mzient — APPROVED** [review]

File: docs/conf.py [hunk 4]
```diff
@@ -563,6 +566,72 @@ def add_directive_header(self, sig):
...
+                break
+
+            # Skip any already valid reference of different kind, for example :func:`foo`
+            match_ref = re.search(r"(:\w+:)(`[^\d\W]\w*`)", line)
+            if (
+                match_ref
+                and m.start() == match_ref.start(2)
+                and m.end() == match_ref.end(2)
+            ):
+                print(f"Skipping in {name}: {line[: m.end()]=}")
```
Comment (klecki) [inline docs/conf.py:589]:
```suggestion
```

File: docs/conf.py [hunk 5]
```diff
@@ -563,6 +566,72 @@ def add_directive_header(self, sig):
...
+                    )
+                result += line[: match_double.end()]
+                line = line[match_double.end() :]
+                continue
+
+            # If we are indeed a parameter, add the :paramref: that sphinx_paramlinks will handle
+            if candidate in params:
+                print(
+                    f"Injecting :paramref: in {name} for {candidate} in {result}{line}"
+                )
```
Comment (klecki) [inline docs/conf.py:617]:
```suggestion
```

**Comment by klecki** [comment]
!build

**Pull request merged**